### PR TITLE
Proposer builds block in parallel. (Consensus vs Execution)

### PR DESCRIPTION
**Pull request by terencechain**
Supersedes #12251

This pull request adds the functionality of constructing consensus and execution parts of blocks in parallel using work groups. The goal is for the process of constructing blocks to be faster and more efficient. By utilizing work groups, different parts of the block can be constructed simultaneously, significantly reducing processing time

### CODE REVIEW

**Comment by prestonvanloon** [comment]
Could you put this into a new method and use a feature flag? Seems like a risky change to not be able to opt out.

**Comment by prestonvanloon** [comment]
Could you add any benchmarks or data to help us understand how significant of a reduction in the processing time?

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go [hunk 1]
```diff
@@ -82,50 +83,62 @@ func (vs *Server) GetBeaconBlock(ctx context.Context, req *ethpb.BlockRequest) (
...
 
-	// Set execution data. New in Bellatrix.
-	if err := vs.setExecutionData(ctx, sBlk, head); err != nil {
+	wg.Add(1)
+	if err := func() error {
+		defer wg.Done()
+		return vs.setExecutionData(ctx, sBlk, head)
+	}(); err != nil {
 		return nil, status.Errorf(codes.Internal, "Could not set execution data: %v", err)
 	}
```
Comment (prestonvanloon) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:139]: This one isn't in a go routine, it's already blocking wg.Wait() so you don't need to use wg here.

```suggestion
	if err := vs.setExecutionData(ctx, sBlk, head); err != nil {
		return nil, status.Errorf(codes.Internal, "Could not set execution data: %v", err)
	}
```

**Review by prestonvanloon — CHANGES_REQUESTED** [review]
Requesting changes until we have more data to support the claim of significant reduction in block creation. #12302 will help with this data collection for builder routes.

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go (hunk 1 above)
Comment (terencechain) [reply beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:139]: Even better. Thank you!

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go [hunk 2]
```diff
@@ -82,50 +85,56 @@ func (vs *Server) GetBeaconBlock(ctx context.Context, req *ethpb.BlockRequest) (
...
-		sBlk.SetDeposits(deposits)
-		sBlk.SetAttestations(atts)
-	}
-
 	// Set proposer index.
 	idx, err := helpers.BeaconProposerIndex(ctx, head)
 	if err != nil {
 		return nil, fmt.Errorf("could not calculate proposer index %v", err)
 	}
 	sBlk.SetProposerIndex(idx)
```
Comment (terencechain) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:93]: Note to the reviewer. The position of setting proposer index has moved up. It was originally after setting attestation, now it's moved up to after setting parent root. It makes the refactoring cleaner

**Comment by prestonvanloon** [comment]
Since this PR, we've added more data collection to our cluster to show that setExecutionData takes 200ms when using builder. The other operations of the block construction consist of about 110ms worth of work. Therefore, this PR should give block construction an average savings of approximately 100ms when using builder route. When using a local block construction, setExecutionData takes only 14ms due to eager caching from the EL. Could we apply some form of eager caching from the builder?

**Comment by terencechain** [comment]
> 

It is the earliest we can get the header, anything earlier will probably result in an error from the relayer

**Pull request merged**